Bundled Cassandra into a fat jar on the classpath

jroper opened the pull request:
This splits the CassandraLauncher into its own artifact. That artifact bundles a Cassandra fat jar, which means that Cassandra and its dependencies do not need to be on the classpath.  When Cassandra is forked, that jar is extracted, and used as the classpath for Cassandra.

This solves the problem of Cassandra (and its many dependencies) needing to be on the classpath, and the conflicts that can create (such as with Netty, Guice, etc).

One issue with this is that it means you can't just add something to your classpath to get Cassandra to use it, such as a custom log configuration, etc. To solve this, an additional classpath parameter has been added to the start method, which allows additional classpath elements to be added to the classpath. A helper has also been added for generating this parameter, that takes a list of resources (eg, logback-test.xml), and calculates the directories/jars that they come from, and returns those directories for adding to the classpath.

raboof commented on build.sbt:
```diff
@@ -92,3 +93,24 @@ osgiSettings
+    name := "akka-persistence-cassandra-bundle",
+    crossPaths := false,
+    autoScalaLibrary := false,
+    libraryDependencies += "org.apache.cassandra" % "cassandra-all" % "3.10" exclude("commons-logging", "commons-logging"),
```
don't we need to exclude netty artifacts here anymore?

jroper replied:
No we don't, because we just produce one fat jar with all the dependencies inside it, and that jar is included as a resource inside the launcher jar, so its classpath is completely separated from and unrelated to whatever is using it.

raboof commented on cassandra-launcher/src/main/scala/akka/persistence/cassandra/testkit/CassandraLauncher.scala:
```diff
@@ -118,50 +172,16 @@ object CassandraLauncher {
-  }
-
-  private def startForked(cassandraDirectory: File, configResource: String, realPort: Int): Unit = {
+  private def startForked(cassandraDirectory: File, configFile: File, cassandraBundle: File, classpath: immutable.Seq[String], port: Int): Unit = {
```
Ah the fact that we always fork and no longer embed is the reason we don't have to exclude netty anymore, then?

Indeed seems neater, are there no use cases for embedding anymore?

jroper replied:
I don't think so - embedding doesn't work on Windows (because Cassandra always calls System.exit when Cassandra shuts down on Windows), and it can't be restarted after it's been started.

jroper commented:
Build is passing now.

I made a change, I added a logback and log4j configuration to the bundle, since the users classpath is no longer used unless they explicitly add elements of it. In most cases, the only thing users will want is to specify a logging configuration, so having a sensible default log configuration made sense. This can easily be overridden if the user needs to.

jroper commented:
This has now been fairly thoroughly tested in https://github.com/lagom/lagom/pull/688 (currently working off a SNAPSHOT dependency).

patriknw commented on cassandra-launcher/src/main/scala/akka/persistence/cassandra/testkit/CassandraLauncher.scala:
```diff
@@ -80,6 +79,21 @@ object CassandraLauncher {
   /**
+   * Use this to add classpath directories to the classpath
+   */
+  def classpathForResources(resources: String*): immutable.Seq[String] = {
```
`@varargs`, for Java API

patriknw commented on cassandra-launcher/src/main/scala/akka/persistence/cassandra/testkit/CassandraLauncher.scala:
```diff
@@ -80,6 +79,21 @@ object CassandraLauncher {
   }
 
   /**
+   * Use this to add classpath directories to the classpath
```
I guess this is used if custom logging or configResource yaml is neeeded. Please add that to this as examples.

patriknw commented:
@raboof CassandraLauncher is s test utility so I think it's fine to move it to separate artifact without worry too much about breaking user code

patriknw commented on build.sbt:
```diff
@@ -2,35 +2,56 @@ import Tests._
+lazy val root = (project in file("."))
+  .enablePlugins(AutomateHeaderPlugin, SbtOsgi)
+  .dependsOn(cassandraLauncher % Optional)
+  .aggregate(cassandraLauncher)
```
so the core project is still defined as "top level", hm, I find that confusing, could you move that to a separate  project?

please also verify that everything works with the release plugin

patriknw commented:
The readme should mention the new artifact and also add a section in the migrations.

jroper commented:
PR has been updated as requested:

* Updated scaladoc of `classpathForResources` to be more clear on its purpose and add examples of why you would use it.
* Added `@varargs` annotation to `classpathForResources`.
* Restructured build such that there is now a dummy root project, and the core library lives under that.
* Added launcher changes to migrations in `README.md` and mentioned it in the docs on how to add it to you `build.sbt`.

patriknw commented:
looking good, the ssl related tests seems to be missing some test resources that should be included in classpath

patriknw commented:
@jroper I fixed the ssl tests, it was the core directory. Merged in https://github.com/akka/akka-persistence-cassandra/pull/202

One thing looks weird when running tests, console output (from scalatest) is not printed until after the test is done. Try 
```
testOnly akka.persistence.cassandra.journal.CassandraJournalPerfSpec
```

patriknw commented:
@jroper released 0.52

jroper commented:
That console output issue also happens with JUnit I think - it might be an sbt thing, and a consequence of not forking?
